docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,500 @@
+
+### TraceWriter
+
+`Newtonsoft.Json` lets you debug by using a `TraceWriter` to view logs that are generated by serialization or deserialization. <xref:System.Text.Json> doesn't do logging.
+
+## JsonDocument and JsonElement compared to JToken (like JObject, JArray)
+
+<xref:System.Text.Json.JsonDocument?displayProperty=fullName> provides the ability to parse and build a **read-only** Document Object Model (DOM) from existing JSON payloads. The DOM provides random access to data in a JSON payload. The JSON elements that compose the payload can be accessed via the <xref:System.Text.Json.JsonElement> type. The `JsonElement` type provides APIs to convert JSON text to common .NET types. `JsonDocument` exposes a <xref:System.Text.Json.JsonDocument.RootElement> property.
+
+### IDisposable
+
+`JsonDocument` builds an in-memory view of the data into a pooled buffer. Therefore, unlike `JObject` or `JArray` from `Newtonsoft.Json`, the `JsonDocument` type implements `IDisposable` and needs to be used inside a using block.
+
+### Read-only
+
+The <xref:System.Text.Json> DOM can't add, remove, or modify JSON elements. It's designed this way for performance and to reduce allocations for parsing common JSON payload sizes (that is, < 1 MB). If your scenario currently uses a writable/modifiable DOM, one of the following workarounds might be feasible:
+
+* To build a `JsonDocument` from scratch (that is, without passing in an existing JSON payload to the `Parse` method), write the JSON text by using the `Utf8JsonWriter` and parse the output from that to make a new `JsonDocument`.
+* To modify an existing `JsonDocument`, use it to write JSON text, making changes while you write, and parse the output from that to make a new `JsonDocument`.
+
+### JsonElement
+
+`JsonDocument` exposes the `RootElement` as a property of type <xref:System.Text.Json.JsonElement>, which is a union, struct type that encompasses any JSON element. `Newtonsoft.Json` uses dedicated hierarchical types like `JObject`,`JArray`, `JToken`, and so forth. `JsonElement` is what you can search and enumerate over, and you can use `JsonElement` to materialize JSON elements into .NET types.
+
+### Search a JsonDocument
+
+Searches for JSON tokens using `JObject` or `JArray` from `Newtonsoft.Json` tend to be relatively fast because they're look-ups in some dictionary. By comparison, searches on `JsonElement` require a sequential search of the properties and hence is relatively slow (for example when using `TryGetProperty`). <xref:System.Text.Json> is designed to minimize initial parse time rather than look-up time. Therefore, use the following approaches to optimize performance when searching through a `JsonDocument` object:
```

Comment:
>generally speaking, APIs should accept `JsonElement` (not `JsonDocument`) 

Is that because JsonDocument is IDisposable or some other reason?

> and you should also generally only return a `JsonElement` that is a sub-element of the passed in `JsonElement`. 

What is the reason for this recommendation?

Reply:
> Is that because JsonDocument is IDisposable

Yes, but also, having the called method do `doc.RootElement` and process/search the whole document is most likely not what you want. The caller should instead pass in the sub-element that is scoped/relevant to the called method.

Otherwise, each API that is accepting a `JsonDocument` would first look for/get the subset they care about.

Of course, scenarios might exists where users may want to pass in `JsonDocument` (and there is nothing wrong with doing so, if needed), such as passing ownership. I just think it wouldn't be very common. And if the method needs to process the whole payload, passing in `JsonDocument.RootElement` would do the trick anyway.

> What is the reason for this recommendation?

There is no way to build up a `JsonElement` and it is immutable. So, if your API is returning a `JsonElement`, it must have got it from somewhere. I can see two such ways:
1) The API got a `JsonElement` as input - it can't modify the contents, but can search within it and return a sub-element (say the `JsonElement` contains this JSON object and the API wants to validate correctness and return the nested JSON array "{foo:[1,2,3]"}).
2) The API used some input source like stream/string/etc. or created a stream/string/etc. and called `JsonDocument.Parse` on it to generate a `JsonDocument`. The API should then either return that `JsonDocument` (to let the caller own it), OR it should `Clone` the `RootElement` (which is a `JsonElement`) before returning it (or a subset of it).

So, maybe we can re-word it to explain the two scenarios, so it is clearer.

That said, again, this is just guideline (because that's what scenarios would require and might be useful for folks to know). Someone could also have an API do combination of both or something else, but I haven't seen such things in practice too often.

```C#
// Some arbitrary counter-example to the common usages of JsonElement within APIs
// LookAndLoad isn't returning a subset of the passed in JsonElement.
public JsonElement LookAndLoad(JsonElement source)
{
   // The passed in JsonElement contains the filename used to load up some JSON
   string json = File.ReadAllText(source.GetProperty("fileName").GetString());
   
   using (JsonDocument doc = JsonDocument.Parse(json))
   {
      return doc.RootElement.Clone();
   }
}
```